### PR TITLE
(feat) O3-3365: Add the ability to record non-coded cause-of-death

### PR DESCRIPTION
**Pull request by jwnasambu**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [ ] My work includes tests or is validated by existing tests.

## Summary
I included a feature to allow users to enter a free-text cause of death. If someone selects 'Other', the text input field will be available for entering this free-text concept

## Screenshots

https://github.com/openmrs/openmrs-esm-patient-chart/assets/33891016/0c30d6a9-bb52-4c1b-8f78-e1c7b9e932a4

## Related Issue
https://openmrs.atlassian.net/browse/O3-3365

### CODE REVIEW

**Comment by jwnasambu** [comment]
Hey @ibacher and @denniskigen, I kindly request for your guidance please! I'm a bit stuck on how to properly utilize `other `cause of death concept fetched by the useCauseOfDeath hook. When I add the TextInput input, it creates another `other ` concept that functions as expected but on submitting the  entered value in the TextInput throws an error as reflected in the video above.

File: packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx [hunk 1]
```diff
@@ -34,25 +35,34 @@ const MarkPatientDeceasedForm: React.FC<DefaultPatientWorkspaceProps> = ({ close
   const { deathDate, isDead } = usePatientDeceasedStatus(patientUuid);
   const [isSubmitting, setIsSubmitting] = useState(false);
   const [searchTerm, setSearchTerm] = useState('');
+  const [showOtherInputField, setShowOtherInputField] = useState(false);
+
+  const OTHER_CONCEPT_UUID = 'UUID of other';
+  const OTHER_CONCEPT_DISPLAY = t('other', 'Other');
```
Comment (denniskigen) [inline packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx:41]:
```suggestion
```suggestion
  const freetextCauseOfDeathUuid = 5622AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA';
```
```

File: packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx [hunk 2]
```diff
@@ -34,25 +35,34 @@ const MarkPatientDeceasedForm: React.FC<DefaultPatientWorkspaceProps> = ({ close
   const { deathDate, isDead } = usePatientDeceasedStatus(patientUuid);
   const [isSubmitting, setIsSubmitting] = useState(false);
   const [searchTerm, setSearchTerm] = useState('');
+  const [showOtherInputField, setShowOtherInputField] = useState(false);
```
Comment (denniskigen) [inline packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx:38]:
```suggestion
```

File: packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx [hunk 3]
```diff
@@ -34,25 +35,34 @@ const MarkPatientDeceasedForm: React.FC<DefaultPatientWorkspaceProps> = ({ close
...
-            extract: (causeOfDeathConcept) => causeOfDeathConcept.display,
-          })
-          .sort((r1, r2) => r1.score - r2.score)
-          .map((result) => result.original)
-      : causesOfDeath;
-  }, [searchTerm, causesOfDeath]);
+    filtered.push({ uuid: OTHER_CONCEPT_UUID, display: OTHER_CONCEPT_DISPLAY, name: OTHER_CONCEPT_DISPLAY });
+
+    return filtered;
+  }, [searchTerm, causesOfDeath, OTHER_CONCEPT_DISPLAY]);
```
Comment (denniskigen) [inline packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx:60]: You don't need any of these changes.

File: packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx [hunk 4]
```diff
@@ -70,13 +81,26 @@ const MarkPatientDeceasedForm: React.FC<DefaultPatientWorkspaceProps> = ({ close
...
+
+  useEffect(() => {
+    setShowOtherInputField(causeOfDeathValue === OTHER_CONCEPT_UUID);
+  }, [causeOfDeathValue]);
+
+  useEffect(() => {
+    if (causesOfDeath && causesOfDeath.length > 0) {
+      setShowOtherInputField(causeOfDeathValue === OTHER_CONCEPT_UUID);
+    }
+  }, [causesOfDeath, causeOfDeathValue]);
```
Comment (denniskigen) [inline packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx:98]: Remove this.

File: packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx [hunk 5]
```diff
@@ -96,7 +120,9 @@ const MarkPatientDeceasedForm: React.FC<DefaultPatientWorkspaceProps> = ({ close
         return;
       }
 
-      markPatientDeceased(deathDate, patientUuid, causeOfDeath)
+      const causeOfDeathToSubmit = causeOfDeath === OTHER_CONCEPT_UUID ? freeTextCauseOfDeath : causeOfDeath;
+
+      markPatientDeceased(deathDate, patientUuid, causeOfDeathToSubmit)
```
Comment (denniskigen) [inline packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx:125]:
```suggestion
      markPatientDeceased(deathDate, patientUuid, causeOfDeath === freetextCauseOfDeathUuid ? freeTextCauseOfDeath : causeOfDeath)
```

File: packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx [hunk 6]
```diff
@@ -192,17 +218,15 @@ const MarkPatientDeceasedForm: React.FC<DefaultPatientWorkspaceProps> = ({ close
...
-                      )}
+                      {filteredCausesOfDeath.map(({ uuid, display, name }) => (
+                        <RadioButton
+                          key={uuid}
+                          className={styles.radioButton}
+                          id={name}
+                          labelText={display}
+                          value={uuid}
+                        />
+                      ))}
```
Comment (denniskigen) [inline packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx:229]: I think we can undo this change as well

File: packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx [hunk 7]
```diff
@@ -216,6 +240,24 @@ const MarkPatientDeceasedForm: React.FC<DefaultPatientWorkspaceProps> = ({ close
...
+                      labelText={t('pleaseSpecify', 'Please specify')}
+                      placeholder={t('enterCauseOfDeath', 'Enter cause of death')}
+                      onChange={onChange}
+                      onBlur={onBlur}
+                      value={value}
+                    />
+                  )}
+                />
+              </div>
+            )}
```
Comment (denniskigen) [inline packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx:260]: Move this below the <section> tag.

File: packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx [hunk 8]
```diff
@@ -216,6 +240,24 @@ const MarkPatientDeceasedForm: React.FC<DefaultPatientWorkspaceProps> = ({ close
               ) : null}
             </div>
             {errors?.causeOfDeath && <p className={styles.errorMessage}>{errors?.causeOfDeath?.message}</p>}
+            {showOtherInputField && (
```
Comment (denniskigen) [inline packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx:243]:
```suggestion
            {causeOfDeathValue === freetextCauseOfDeathUuid && (
```

**Comment by jwnasambu** [comment]
@denniskigen Thanks

**Comment by denniskigen** [comment]
@ibacher what property should be used to record the freetext value to the backend? https://rest.openmrs.org/#create-a-person

**Comment by ibacher** [comment]
`causeOfDeathNonCoded`

**Comment by jwnasambu** [comment]
@denniskigen this is funny. We walking on the same issue at ago. Thanks  am looking at some of the changes you have pushed and I can consent I wasn't thinking in that direction and I will be glad learning something from the coffee break session.

**Comment by jwnasambu** [comment]
@denniskigen I was concentrating on validating this form since it can be submitted successfully with or without a date. Should I continue with the task or terminate it the task?

**Comment by jwnasambu** [comment]
@denniskigen am so sorry for the late response I had issues with internet.  Kindly I have tested and attached the current snapshot  everything seems to be working as expected.

**Review by denniskigen — APPROVED** [review]
LGTM

File: packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx [hunk 9]
```diff
@@ -34,12 +35,12 @@ const MarkPatientDeceasedForm: React.FC<DefaultPatientWorkspaceProps> = ({ close
   const { deathDate, isDead } = usePatientDeceasedStatus(patientUuid);
   const [isSubmitting, setIsSubmitting] = useState(false);
   const [searchTerm, setSearchTerm] = useState('');
+  const freetextCauseOfDeathUuid = '5622AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA';
```
[the remark this answers is not in the document]
Comment (ibacher) [reply packages/esm-patient-chart-app/src/mark-patient-deceased/mark-patient-deceased-form.workspace.tsx]: Yes. Just about anywhere we use a concept, the concept should be configurable. Technically, the platform only enforces that there is a True / False concept.
